File: PyVersion/metrics.py
from __future__ import annotations
from dataclasses import dataclass
from functools import cached_property
from typing import Dict, List
import math, time, random   

from models import Anchor, Tag, PathLossModel
from utils  import _R3_distance, _logpdf_student_t, _cep95_from_conf

EWMA_THRESHOLD: float = 8.0

def update_anchors_from_tag_data(
    anchors: List[Anchor], 
    tag: Tag, 
    model: PathLossModel, 
    now: float, 
    deltaR: float = 7.0, #de base: 10.0
    T_vis: int = 6000
) -> None:
    """
    Updates the shared state of all anchors based on data from a single tag.
    This includes both path-loss parameters (RSSI0, n) and health (ewma).
    """
    temp_system = TagSystem(tag, model)

    # --- Parameter updates (from old update_anchor_parameters) ---
    for anchor in temp_system._get_significant_anchors(anchors):
        if anchor.macadress in tag.rssi_dict:
            rssi = tag.rssi_for_anchor(anchor.macadress)
            dist = _R3_distance(anchor.coord, tag.est_coord)
            anchor.update_parameters(rssi, dist)

    # --- Health updates (from old update_anchor_healths) ---
    if not tag.rssi_dict:
        return

    # To calculate z-values for health, we need a temporary system
    temp_system = TagSystem(tag, model)
    z_dict = temp_system.z_vals(anchors) # Pass anchors in

    max_rssi = max(tag.rssi_dict.values())
    anchor_map = {a.macadress: a for a in anchors}

    for anchor_id, z_val in z_dict.items():
        anchor = anchor_map[anchor_id] 
        anch_rssi = tag.rssi_for_anchor(anchor_id)
        delta_rssi = max_rssi - anch_rssi
        time_since = now - anchor.last_seen if anchor.last_seen else 0.0

        # Gate checks
        if delta_rssi > deltaR: 
            continue
        if time_since > T_vis:
            continue 

        anchor.update_health(z_val, now)

@dataclass(slots=False)
class TagSystem:
    tag: Tag
    model: PathLossModel

    def _get_significant_anchors(self, anchors: List[Anchor], max_n: int = 5) -> List[Anchor]:
        rssi_dict = self.tag.rssi_dict
        if not rssi_dict:
            return []
        max_rssi = max(rssi_dict.values())
        keep = [
            a for a in anchors
            if a.macadress in rssi_dict and rssi_dict[a.macadress] >= max_rssi - 10 and a.ewma < EWMA_THRESHOLD
        ]
        keep.sort(key=lambda a: rssi_dict[a.macadress], reverse=True)
        return keep[:max_n]

    def distances(self, anchors: List[Anchor]) -> Dict[str, float]:
        return {
            anchor.macadress: _R3_distance(anchor.coord, self.tag.est_coord)
            for anchor in self._get_significant_anchors(anchors)
        }   
    
    def z_vals(self, anchors: List[Anchor]) -> Dict[str, float]:
        # Only use significant anchors
        significant_anchors = self._get_significant_anchors(anchors)
        dist_dict = {
            anchor.macadress: _R3_distance(anchor.coord, self.tag.est_coord)
            for anchor in significant_anchors
        }
        z_values = {}
        anchor_map = {a.macadress: a for a in anchors}
        for anch_id, est_dist in dist_dict.items():
            anchor = anchor_map[anch_id]
            rssi_value = self.tag.rssi_for_anchor(anch_id)
            z_values[anch_id] = self.model.z(rssi_value, anchor.RSSI_0, anchor.n, est_dist)
        return z_values
    
    def confidence_score(self, anchors: List[Anchor], v: int = 5, scale: float = 2) -> float:
        z_dict = self.z_vals(anchors)
        if not z_dict:
            return 0.0        
        sig = sum(_logpdf_student_t(z, v) for _, z in z_dict.items())
        l = sig / len(z_dict)
        return math.exp(l / scale)

    def error_radius(self, anchors: List[Anchor]) -> float:
        P_val = self.confidence_score(anchors)
        return _cep95_from_conf(P_val)

File: CppVersion/metrics.cpp
#include <algorithm>    
#include <cmath>        
#include <numeric>
#include <vector>
#include <unordered_map>
#include <string>
#include <limits>

#include "metrics.h"
#include "config.h"

/*TAGSYSTEM*/
//constructor:
TagSystem::TagSystem(const Tag& inpt_tag, const PathLossModel& inpt_model) : tag(inpt_tag), model(inpt_model) {
}

//getters:
const Tag& TagSystem::get_tag() const {
    return tag;
}

const PathLossModel& TagSystem::get_model() const {
    return model;
}

//methods
std::vector<Anchor*> TagSystem::get_significant_anchors(std::vector<Anchor*>& anch_list, int max_n) {
    const std::unordered_map<std::string, float>& rssi_dict = tag.get_rssi_readings();
    
    if (rssi_dict.empty()) {
        return {};
    }
    
    float max_rssi = std::numeric_limits<float>::lowest();
    for (const auto& [key, value] : rssi_dict) {
        if (value > max_rssi) {
            max_rssi = value;
        }
    }

    std::vector<Anchor*> keep;
    
    for (auto* anchor : anch_list) {
        std::string anchor_mac = anchor->get_mac_address();
        auto rssi_it = rssi_dict.find(anchor_mac);
        
        if (rssi_it != rssi_dict.end() && 
            rssi_it->second >= (max_rssi - 10) && 
            anchor->get_ewma() < Calibration::EWMA_THRESHOLD) {
            keep.push_back(anchor);
        }
    }

    std::sort(keep.begin(), keep.end(),
        [&rssi_dict](const Anchor* a1, const Anchor* a2) {
            return rssi_dict.at(a1->get_mac_address()) > rssi_dict.at(a2->get_mac_address());
        }
    );

    // Truncate to top max_n
    if (keep.size() > static_cast<size_t>(max_n)) {
        keep.resize(max_n);
    }
    return keep;
}

std::unordered_map<Anchor*, float> TagSystem::distances(std::vector<Anchor*>& anch_list) {
    std::unordered_map<Anchor*, float> result;
    std::vector<Anchor*> significant_anchors = get_significant_anchors(anch_list);

    for (const auto& anchor : significant_anchors) {
        float dist = R3_distance(anchor->get_coord(), tag.get_est_coord());
        result[anchor] = dist;
    }
    return result;
}

std::unordered_map<Anchor*, float> TagSystem::z_vals(std::vector<Anchor*>& anch_list) {
    std::unordered_map<Anchor*, float> result;
    std::unordered_map<Anchor*, float> dist_dict = distances(anch_list);
    const std::unordered_map<std::string, float>& rssi_dict = tag.get_rssi_readings();

    for (const auto& pair : dist_dict) {
        float rssi_value = rssi_dict.at(pair.first->get_mac_address());
        result[pair.first] = model.z(rssi_value, pair.first->get_RSSI_0(), pair.first->get_n(), pair.second);
    }
    
    return result;
}

float TagSystem::confidence_score(std::vector<Anchor*>& anch_list, int v, float scale) {
    std::unordered_map<Anchor*, float> z_dict = z_vals(anch_list);
    if (z_dict.empty()) {
        return 0.0;
    }

    float weighted_sig = 0.0f;
    float total_weight = 0.0f;

    for (const auto& [anchor, z_val] : z_dict){    
        float anchor_ewma = anchor->get_ewma();
        float anchor_weight = 1.0f / (1.0f + anchor_ewma + std::pow(z_val, 2));
        weighted_sig += anchor_weight * logpdf_student_t(z_val, v);
        total_weight += anchor_weight;
    }

    float l = weighted_sig / total_weight;
    return std::exp(l / scale);
}

float TagSystem::error_radius(std::vector<Anchor*>& anch_list) {
    float p_val = confidence_score(anch_list);
    return cep95_from_conf(p_val);
}


/*METHOD*/
void update_anchors_from_tag_data(
    std::vector<Anchor*>& anch_list, 
    const Tag& inpt_tag, 
    const PathLossModel& inpt_model, 
    float now, 
    float deltaR, 
    int T_vis
){
    TagSystem moment_system = TagSystem(inpt_tag, inpt_model);
    const std::unordered_map<std::string, float>& rssi_dict = moment_system.get_tag().get_rssi_readings();

    //paramaters update
    if (rssi_dict.empty()) {
        return;
    }
    
    // Parameter updates - work directly with pointers to original anchors
    std::vector<Anchor*> significant_anchors = moment_system.get_significant_anchors(anch_list);
    std::unordered_map<Anchor*, float> distance_dict = moment_system.distances(anch_list);
    
    for (const auto& sign_anchor : significant_anchors) {
        float sign_anchor_rssi = rssi_dict.at(sign_anchor->get_mac_address());
        float sign_anchor_dist = distance_dict.at(sign_anchor);
        sign_anchor->update_parameters(sign_anchor_rssi, sign_anchor_dist);
    }

    //health update
    std::unordered_map<Anchor*, float> z_dict = moment_system.z_vals(anch_list);
    
    float max_rssi = std::numeric_limits<float>::lowest();
    for (const auto& pair : rssi_dict) {
        if (pair.second > max_rssi) {
            max_rssi = pair.second;
        }
    }

    for (const auto& pair : z_dict) {
        Anchor* sign_anchor = pair.first;
        float sign_anchor_rssi = rssi_dict.at(sign_anchor->get_mac_address());
        float rssi_delta = max_rssi - sign_anchor_rssi;
        float sign_anchor_z_val = pair.second;

        float time_since_last_seen = 0.0;
        if (sign_anchor->get_last_seen() != 0.0) {
            time_since_last_seen = now - sign_anchor->get_last_seen();
        }

        if (time_since_last_seen > T_vis || rssi_delta > deltaR) {
            continue;
        }
        
        // Directly update the anchor via pointer - no find_if needed!
        sign_anchor->update_health(sign_anchor_z_val, now);
    }
}